### ISSUE TITLE
Why require node18?

### ISSUE THREAD

**Issue by flchaux** [derhuerst/email-providers#15]
Hi, 

Thank you for this package. We just saw you are requiring node18 in the last version (2.0.0). We don't understand, it's just static files, isn't it?

Florian.

**Comment by derhuerst** [comment]
Yes, `email-providers` contains just static JSON.

The development dependencies require a more recent Node.js version than before (`6`), e.g. `node-fetch@3` is ESM-only. Because I don't want to claim support for platforms I can't properly `build` & `test` on, I raised the minimum Node.js version to 18. I would say that this shouldn't be a big problem, given that Node.js v18 will become (active) LTS in a few days and that Node v16 will go out of (active) LTS.

If you want to have up-to-date data on Node <18, use `email-providers@1.14.0` which I have published 10 days ago.

**Comment by santialbo** [comment]
> Because I don't want to claim support for platforms I can't properly build & test on, I raised the minimum Node.js version to 18.

One thing is building the library, another one is consuming it. I don't think this package should be forcing node 18 on consumers. Specially given that the output is just a json file.

**Comment by derhuerst** [comment]
(Sorry to be salty here, but I feel like I have explained the reasoning already.)

If it's just a JSON file, you can pull it from https://unpkg.com or copy it into place?

I actually think the appeal of this package is the (rather small) service on my side of updating the data every now and then. You're benefitting from this service already.

</sarcasm>

I haven't updated the package regularly *anyways* in the past years, so what's the problem with using `email-providers@1` if you use Node 16? I could do another update in 6 months.

**Comment by Jule-** [comment]
@derhuerst thank you for this useful package. I don't think @santialbo wanted to be disrespectful. But I think he made a point and even if I understand your point of view and the fact that you are not alone with the same issue regarding the lack of `devEngine` field in `package.json` I think the warning when consuming your package is misleading.

I feel like a prerequisite section in the `README.md` should do the trick to clarify what node versions you want to claim support on. Additionally, even if not the rule, you can take benefit of the fact that `nvm` is widely used to add a `.nvmrc` file to your project. It could free you from the hassle of maintaining the `v1.x.x` (which is not evident to trust for someone consuming your package).

That being said, ESM is not only available in v18 (https://nodejs.org/api/esm.html#modules-ecmascript-modules), maybe you could lower the requirement, at least to v16, which is widely used.
Even `node-fetch` have a such requirement:
https://github.com/node-fetch/node-fetch/blob/8ced5b941cf36d0d7e0c1017aa2a4abcb29ecd89/package.json#L13-L15
```json
  "engines": {
    "node": "^12.20.0 || ^14.13.1 || >=16.0.0"
  },
```

No matter what, thank you for your work and I will respect your choice, just wanted to share my thoughts. 😉

**Comment by derhuerst** [comment]
> I think the warning when consuming your package is misleading.

I'll answer in a very generalised way. I think you're misinterpreting what I intend to express with `engines.node: >= 18`:

Because you're consuming my open source library, you're trusting my *promise* to keep it working under specific conditions (Node >= 18). The `engines` field merely documents my intention to keep it compatible in a machine-readable way.

Obviously, you cannot blindly trust this *promise* 100%, especially not in critical and/or sensitive environments: You'll have to test and/or vet every `email-providers` version anyways if you want to be *sure* it's working as intended.

When you want to use `email-providers` under different conditions than those I promise compatibility for, you have three options:
- decide to take the risk (in this case, as pointed out, there is almost no risk, because `email-providers` is a static JSON file) – In this case, live with the warning that npm prints, or try to configure it not to show it.
- offer support to maintain a version of `email-providers` that supports Node <18.
- fork the project. (I see no reason to do this, I'm perfectly fine making you collaborators.)

The whole discussion seems a bit superfluous to me though, since you apparently feel a strong need to use the latest version of `email-providers`, even though we're talking about a package that I have rarely updated in the past years, so you haven't had up-to-date data *anyways*.

> I feel like a prerequisite section in the `README.md` should do the trick to clarify what node versions you want to claim support on.

Isn't that exactly what `engines.node` does? [The `engines` documentation](https://docs.npmjs.com/cli/v9/configuring-npm/package-json#engines) says:

> You can specify the version of node that your stuff works on [...]
> Unless the user has set the [engine-strict config](https://docs.npmjs.com/cli/v9/using-npm/config#engine-strict) flag, this field is advisory only and will only produce warnings when your package is installed as a dependency.

> Additionally, even if not the rule, you can take benefit of the fact that `nvm` is widely used to add a `.nvmrc` file to your project. It could free you from the hassle of maintaining the `v1.x.x` (which is not evident to trust for someone consuming your package).

How does that help me? I converted the build & test scripts to ESM, so I needed Node >= 16 (see below), and I didn't want to claim support for Node versions that I don't run tests with. (And yes, as elaborated, it is just a plain JSON file we're testing.) Using the right Node version is not a problem here I think.

> That being said, ESM is not only available in v18 (https://nodejs.org/api/esm.html#modules-ecmascript-modules), maybe you could lower the requirement, at least to v16, which is widely used.

**Indeed, I missed that Node 16 supports it. We can change it to `>= 16` if you want.**

> Even `node-fetch` have a such requirement [...].

Sorry to be insulting here, but to put it bluntly for the sake of getting my point across, your statement really fits the cliché of an entitled open source consumer: You use my free (in both senses) work and complain that I don't provide the same guarantees as a [library maintained by a (I assume paid) team](https://github.com/node-fetch/node-fetch/blob/8ced5b941cf36d0d7e0c1017aa2a4abcb29ecd89/README.md#team).

**Comment by Jule-** [comment]
Sorry, I didn't want to revive a big discussion about this. I just wanted to try to give you another point of view, I thought it helpful, but it seems not.

> The whole discussion seems a bit superfluous to me though, since you apparently feel a strong need to use the latest version of email-providers, even though we're talking about a package that I have rarely updated in the past years, so you haven't had up-to-date data anyways.

You have to understand that when you use a package for the first time, you do not know all its history. Are you suggesting new consumers should read all issues, PR, and versions to understand that they could use an anterior version for other purposes and come back often to see if any improvements/security fixes/... on the last major version are not fed back to the old ones? That is why most developers are trying to keep their dependencies up-to-date, and this one is one amongst the others, nothing related to the freshness of your list.

> Isn't that exactly what `engines.node` does? [The `engines` documentation](https://docs.npmjs.com/cli/v9/configuring-npm/package-json#engines) says:

Yes, you are right. But I thought that the _promise_ or the _value proposition_ of this package, as published on npm, was a curated list of email providers as JSON files (JS native). Which is perfect IMHO. While the backstage of the creation and the quality of these files are handled by your `build.js` and `test.js` files, which are the ones that require specific Node versions, and potentially the `index.js` which uses `import.meta.url` (seems introduced in v10.4.0). Unless you plan to change something in your package or if I have missed something, this node version **for consumers** is unnecessarily restrictive.
You have the right to not even care and state that you do not want to bother with old versions if you make any changes to this version. This is why you put this requirement, as I understand from your last comment. Honestly, I am not here to give lessons, and as you said, everyone can fork etc... Just feels like you are the right person to keep this package up and all the credit. Plus, as you also mentioned, this is a warning, and consumers could decide to live with it until upgrading their node environment to v18, so not a blocking issue.

> How does that help me? I converted the build & test scripts to ESM, so I needed Node >= 16 (see below), and I didn't want to claim support for Node versions that I don't run tests with. (And yes, as elaborated, it is just a plain JSON file we're testing.) Using the right Node version is not a problem here I think.

It was because I thought, in the first place, you wanted to tell potential contributors to use the right Node version for the build and test scripts.

> Sorry to be insulting here, but to put it bluntly for the sake of getting my point across, your statement really fits the cliché of an entitled open source consumer: You use my free (in both senses) work and complain that I don't provide the same guarantees as a [library maintained by a (I assume paid) team](https://github.com/node-fetch/node-fetch/blob/8ced5b941cf36d0d7e0c1017aa2a4abcb29ecd89/README.md#team).

I feel bad you took it like that. I am not native English, and maybe "even" made you think that way, but I was not complaining at all... My point was because you said in a previous comment: `e.g. node-fetch@3 is ESM-only`. And I just wanted to point out the fact that "ESM-only" is not a reason to put `"node": ">=18"`...

I just wanted to let you know that there's no need to be insulting. I was respectful from the beginning. You know, people taking time to give you insights, feedback, and opinions on your project are not "entitled open source consumer"; they are open source **contributors**, stopping their work, taking time, maybe trying hard to be understood in a non-native language, to potentially turn the projects they love, promote and use better for the whole, and they could even pull requests to make your project better if driven to. Your vision of open source sounds a little bit cliché while mine is a little bit idealistic. I let you judge which one serves open source the most.

Never mind, I heard you loud and clear, and I won't make any further comments. I respect your time, you do not need to answer.

Thanks again for your answer and the time you took to expose your vision.

**Comment by derhuerst** [comment]
> You have to understand that when you use a package for the first time, you do not know all its history.

No, not at all!

> Are you suggesting new consumers should read all issues, PR, and versions to understand that they could use an anterior version for other purposes and come back often to see if any improvements/security fixes/... on the last major version are not fed back to the old ones?

If they want to propose a change, and if they know how to "investigate" these things, yes, I think they should check if a topic has been discussed before, or if there are relevant PRs or commits.

> [...] I wanted to say "claim support on" regarding the contribution/building part, not the consuming part.

I guess we could test the built data on more versions than we build them on. I'd be happy to merge a PR that adapts the CI workflow accordingly.

Not sure this is still relevant though because, as you pointed out, we can lower `engines.node` to `>= 16`, and Node 14 will go out of maintenance in 1.5 months.

> Unless you plan to change something in your package or if I have missed something, this node version **for consumers** is unnecessarily restrictive.

As things stand right now, definitely! I would want to make sure it *stays* compatible over time though, so I think there needs to be a CI step testing this. I could be as simple as `npx node@14 -p 'require(".")'`.

I (try to) maintain more than 300 npm packages. I try to reduce maintenance effort and amount of setup specific to each repo to a minimum, in order not to make mistakes when publishing a new version years later.

> > How does that help me? I converted the build & test scripts to ESM, so I needed Node >= 16 (see below), and I didn't want to claim support for Node versions that I don't run tests with. (And yes, as elaborated, it is just a plain JSON file we're testing.) Using the right Node version is not a problem here I think.
> 
> It was because I thought, in the first place, you wanted to tell potential contributors to use the right Node version for the build and test scripts.

We had a misunderstanding then, sorry. I was talking about consumers, but explaining why I raised `engines.node` to `>= 18` (because of building & testing).

> > Sorry to be insulting here, but to put it bluntly for the sake of getting my point across, your statement really fits the cliché of an entitled open source consumer: You use my free (in both senses) work and complain that I don't provide the same guarantees as a [library maintained by a (I assume paid) team](https://github.com/node-fetch/node-fetch/blob/8ced5b941cf36d0d7e0c1017aa2a4abcb29ecd89/README.md#team).
> 
> I feel bad you took it like that. I am not native English, and maybe "even" made you think that way, but I was not complaining at all... My point was because you said in a previous comment: `e.g. node-fetch@3 is ESM-only`. And I just wanted to point out the fact that "ESM-only" is not a reason to put `"node": ">=18"`...

I missed that connection sorry. Because I had misinterpreted your comment, I had been a bit disgruntled already, so I thought you were expressing that you would like me to put in more effort.

Thank you for pointing out that Node 16 works too!

> I just wanted to let you know that there's no need to be insulting. I was respectful from the beginning.

True, I lost my temper a bit.

> You know, people taking time to give you insights, feedback, and opinions on your project are not "entitled open source consumer"; they are open source **contributors**, stopping their work, taking time, maybe trying hard to be understood in a non-native language, to potentially turn the projects they love, promote and use better for the whole, and they could even pull requests to make your project better if driven to.

You might not believe from reading my comments here, but I can relate to this perspective very much! As someone who frequently contributes to other projects, I know that it takes effort and endurance.

As I explained, I reacted this way because I thought you were expecting the same level of maintenance effort as `node-fetch`.

> Your vision of open source sounds a little bit cliché while mine is a little bit idealistic. I let you judge which one serves open source the most.
> 
> Never mind, I heard you loud and clear, and I won't make any further comments. I respect your time, you do not need to answer.

I would like you to make contributions in the future, including raising points via comments!

I think we have discussed this topic sufficiently though, I will a push a commit that lowers `engines.node` to `>= 16`.

**Comment by derhuerst** [comment]
I have published `email-providers@2.0.1`, which requires only `>= 16`.

**Comment by Jule-** [comment]
@derhuerst :pray: really appreciate mate, I understand you all. Happy to cross your way again in my open-source journey. :wink:
